src/EditorFeatures/Core/Implementation/IntelliSense/Completion2/RoslynCompletionItemSource.cs

```diff
@@ -0,0 +1,183 @@
+        {
+            return Task.FromResult<object>("Documentation!"); // Editor doesn't call this yet
+        }
+
+        public void CustomCommit(ITextView view, ITextBuffer buffer, Prototype.CompletionItem item, ITrackingSpan applicableSpan, string textEdit)
+        {
+            // We should crash if this fails
+            var service = GetCompletionService(buffer.CurrentSnapshot) as CompletionServiceWithProviders;
+
+            var roslynItem = item.Properties.GetProperty<Microsoft.CodeAnalysis.Completion.CompletionItem>(RoslynItem); // We're using custom data we deposited in GetCompletionContextAsync
+            var triggerSnapshot = item.Properties.GetProperty<ITextSnapshot>(TriggerSnapshot);
+
+            var edit = buffer.CreateEdit();
+            var provider = service.GetProvider(roslynItem);
+            if (provider is ICustomCommitCompletionProvider c)
+            {
+                c.Commit(roslynItem, view, buffer, triggerSnapshot, null);
+            }
+            else
+            {
+                var document = buffer.GetRelatedDocuments().First();
+                char? commitCharacter = String.IsNullOrEmpty(textEdit) ? null : new char?(textEdit[0]);
+                var roslynChange = service.GetChangeAsync(document, roslynItem, commitCharacter, CancellationToken.None).Result;
+
+                // TODO: Editor to reapply inserted trigger after we commit
+                var ts = new SnapshotSpan(triggerSnapshot, roslynChange.TextChange.Span.ToSpan());
+                var mapped = ts.TranslateTo(buffer.CurrentSnapshot, SpanTrackingMode.EdgeInclusive);
+
+                edit.Replace(mapped.Span, roslynChange.TextChange.NewText);
+            }
+
+            edit.Apply();
+        }
+
+        public ImmutableArray<string> GetPotentialCommitCharacters()
+        {
+            return CommitChars;
+        }
+
+        public bool ShouldCommitCompletion(string typedChar, SnapshotPoint location)
+        {
+            return CommitChars.Contains(typedChar);
+        }
+
+        public bool ShouldTriggerCompletion(string edit, SnapshotPoint location)
+        {
+            var text = SourceText.From(location.Snapshot.GetText());
+            var service = GetCompletionService(location.Snapshot);
+            return service?.ShouldTriggerCompletion(text, location.Position, RoslynTrigger.CreateInsertionTrigger(edit[0])) ?? false;
+        }
+
+        private CompletionService GetCompletionService(ITextSnapshot snapshot)
+        {
+            Document document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return null;
+            }
+
+            var workspace = document.Project.Solution.Workspace;
+            return workspace.Services.GetLanguageServices(LanguageNames.CSharp).GetService<CompletionService>();
+        }
+
+        public Task HandleViewClosedAsync(ITextView view) => throw new NotImplementedException();
```

Comment:
Don't throw, do nothing. This method is used by language services that are not aware of workspace and need to clean up on the fly